src/engine/input/RunOnInputAttribute.cs

```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+public abstract class RunOnInputAttribute : Attribute
+{
+    ~RunOnInputAttribute()
+    {
+        AttributesWithMethods.RemoveAll(p => p.Item2.Equals(this));
+    }
+
+    public static IList<object> InputClasses { get; } = new List<object>();
```

Comment:
I think you actually mean InputReceivingInstances here? And not classes.
Use List here.